diff --git a/MagicalCryptoWallet/Blockchain/Transactions/TransactionFactory.cs b/MagicalCryptoWallet/Blockchain/Transactions/TransactionFactory.cs
index 6189a0bd2f..b2d23acf47 100644
--- a/MagicalCryptoWallet/Blockchain/Transactions/TransactionFactory.cs
+++ b/MagicalCryptoWallet/Blockchain/Transactions/TransactionFactory.cs
@@ -222,8 +222,7 @@ public class TransactionFactory
 
 		// Build the transaction
 
-		psbt.AddKeyPaths(KeyManager);
-		psbt.AddPrevTxs(_transactionStore);
+		psbt = MagicalCryptoWallet.Mcw.Psbt.McwPsbtMetadata.Enrich(psbt, KeyManager, _transactionStore);
 
 		Transaction tx;
 		if (!parameters.TryToSign)
diff --git a/MagicalCryptoWallet/Extensions/NBitcoinExtensions.cs b/MagicalCryptoWallet/Extensions/NBitcoinExtensions.cs
index 8edbcd572f..ca0b36ed90 100644
--- a/MagicalCryptoWallet/Extensions/NBitcoinExtensions.cs
+++ b/MagicalCryptoWallet/Extensions/NBitcoinExtensions.cs
@@ -241,63 +241,6 @@ public static class NBitcoinExtensions
 		return null;
 	}
 
-	/// <summary>
-	/// Tries to equip the PSBT with input and output keypaths on best effort.
-	/// </summary>
-	public static void AddKeyPaths(this PSBT psbt, KeyManager keyManager)
-	{
-		if (keyManager.MasterFingerprint.HasValue)
-		{
-			var fp = keyManager.MasterFingerprint.Value;
-
-			// Add input keypaths.
-			foreach (var script in psbt.Inputs.Select(x => x.WitnessUtxo?.ScriptPubKey).ToArray())
-			{
-				if (script is { })
-				{
-					if (keyManager.TryGetKeyForScriptPubKey(script, out HdPubKey? hdPubKey))
-					{
-						psbt.AddKeyPath(fp, hdPubKey, script);
-					}
-				}
-			}
-
-			// Add output keypaths.
-			foreach (var script in psbt.Outputs.Select(x => x.ScriptPubKey).ToArray())
-			{
-				if (keyManager.TryGetKeyForScriptPubKey(script, out HdPubKey? hdPubKey))
-				{
-					psbt.AddKeyPath(fp, hdPubKey, script);
-				}
-			}
-		}
-	}
-
-	public static void AddKeyPath(this PSBT psbt, HDFingerprint fp, HdPubKey hdPubKey, Script script)
-	{
-		var rootKeyPath = new RootedKeyPath(fp, hdPubKey.FullKeyPath);
-		psbt.AddKeyPath(hdPubKey.PubKey, rootKeyPath, script);
-	}
-
-	/// <summary>
-	/// Tries to equip the PSBT with previous transactions with best effort. Always <see cref="AddKeyPaths"/> first otherwise the prev tx won't be added.
-	/// </summary>
-	public static void AddPrevTxs(this PSBT psbt, ITransactionStore transactionStore)
-	{
-		// Fill out previous transactions.
-		foreach (var psbtInput in psbt.Inputs)
-		{
-			if (transactionStore.TryGetTransaction(psbtInput.PrevOut.Hash, out var tx))
-			{
-				psbtInput.NonWitnessUtxo = tx.Transaction;
-			}
-			else
-			{
-				Logger.LogDebug($"Transaction id: {psbtInput.PrevOut.Hash} is missing from the {nameof(transactionStore)}. Ignoring...");
-			}
-		}
-	}
-
 	public static FeeRate GetSanityFeeRate(this MemPoolInfo me)
 	{
 		var mempoolMinFee = (decimal)me.MemPoolMinFee;
